In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

In [2]:
!pip install wikipedia

  Running setup.py bdist_wheel for wikipedia ... done
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/87/2a/18/4e471fd96d12114d16fe4a446d00c3b38fb9efcb744bd31f4a
Successfully built wikipedia


In [20]:
import requests
wikipedia_link='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
r = requests.get(wikipedia_link)
page = r.text
import lxml.html
htmlNode = lxml.html.fromstring(page)
rawTitle = htmlNode.find(".//title").text
print(rawTitle)

import wikipedia as wp
 
#Get the html source
html = wp.page("List_of_postal_codes_of_Canada:_M").html().encode("UTF-8")
df = pd.read_html(html)[0]
df.columns = df.iloc[0]
df = df.reindex(df.index.drop(0))
df=df[df.Borough != 'Not assigned']
df = df.reset_index(drop=True)
print (df.shape)
df = df.groupby(['Postcode'])['Neighbourhood'].apply(', '.join).reset_index()
df.head()



List of postal codes of Canada: M - Wikipedia
(212, 3)


Postcode                           Neighbourhood
0      M1B                          Rouge, Malvern
1      M1C  Highland Creek, Rouge Hill, Port Union
2      M1E       Guildwood, Morningside, West Hill
3      M1G                                  Woburn
4      M1H                               Cedarbrae

In [ ]:
postalcode = pd.read_csv('https://cocl.us/Geospatial_data')
postalcode.columns = ['Postcode','Latitude','Longitude']
postalcode.head()

result = pd.merge(df,
                 postalcode,left_on='Postcode', right_on='Postcode',
                )
result.head()


#### Load and explore the data

Next, let's load the data.

In [22]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Let's take a quick look at the data.

In [23]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

Notice how all the relevant data is in the *features* key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [24]:
neighborhoods_data = newyork_data['features']

Let's take a look at the first item in this list.

In [25]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Tranform the data into a *pandas* dataframe

The next task is essentially transforming this data of nested Python dictionaries into a *pandas* dataframe. So let's start by creating an empty dataframe.

In [26]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Take a look at the empty dataframe to confirm that the columns are as intended.

In [27]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [28]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [29]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [30]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [31]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7308619, -73.9871558.


#### Create a map of New York with neighborhoods superimposed on top.

In [32]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [33]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

In [34]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7900869, -73.9598295.


As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.

In [35]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [36]:
CLIENT_ID = '4LF1JFD5K2R3VSKEKVVYYVJ0ZAU3XADJMOSVF4MCXO2ZH5AC' # your Foursquare ID
CLIENT_SECRET = '3ITP4YAR5BRPDEMFLEVZGYI4F04A2LI3ITEI1JVPMVQINMBN' # your Foursquare Secret
VERSION = '20180604'


print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4LF1JFD5K2R3VSKEKVVYYVJ0ZAU3XADJMOSVF4MCXO2ZH5AC
CLIENT_SECRET:3ITP4YAR5BRPDEMFLEVZGYI4F04A2LI3ITEI1JVPMVQINMBN


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [37]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

Get the neighborhood's latitude and longitude values.

In [38]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [41]:
# type your answer here


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    500, 
    100)
url # display URL


'https://api.foursquare.com/v2/venues/explore?&client_id=4LF1JFD5K2R3VSKEKVVYYVJ0ZAU3XADJMOSVF4MCXO2ZH5AC&client_secret=3ITP4YAR5BRPDEMFLEVZGYI4F04A2LI3ITEI1JVPMVQINMBN&v=20180604&ll=40.87655077879964,-73.91065965862981&radius=500&limit=100'

Double-click __here__ for the solution.
<!-- The correct answer is:
LIMIT = 100 # limit of number of venues returned by Foursquare API
-->

<!--
radius = 500 # define radius
-->

<!--
\\ # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL
--> 

Send the GET request and examine the resutls

In [42]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c570cdb4434b974106a51d7'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 24,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b4429abf964a52037f225e3',
       'name': "Arturo's",
       'location': {'address': '5198 Broadway',
        'crossStreet': 'at 225th St.',
        'lat': 40.87441177110231,
        'lng': -73.91027100981574,
        'labeledLatLngs': [{'label'

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [43]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [44]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name          categories        lat        lng
0               Arturo's         Pizza Place  40.874412 -73.910271
1            Bikram Yoga         Yoga Studio  40.876844 -73.906204
2          Tibbett Diner               Diner  40.880404 -73.908937
3              Starbucks         Coffee Shop  40.877531 -73.905582
4  Land & Sea Restaurant  Seafood Restaurant  40.877885 -73.905873

And how many venues were returned by Foursquare?

In [45]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

24 venues were returned by Foursquare.


<a id='item2'></a>

## 2. Explore Neighborhoods in Manhattan

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [46]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            500, 
            100)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *manhattan_venues*.

In [48]:
# type your answer here

manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )



Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown


KeyboardInterrupt: 

Double-click __here__ for the solution.
<!-- The correct answer is:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )
--> 

#### Let's check the size of the resulting dataframe

In [49]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3308, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Marble Hill              40.876551               -73.91066   
1  Marble Hill              40.876551               -73.91066   
2  Marble Hill              40.876551               -73.91066   
3  Marble Hill              40.876551               -73.91066   
4  Marble Hill              40.876551               -73.91066   

                   Venue  Venue Latitude  Venue Longitude      Venue Category  
0               Arturo's       40.874412       -73.910271         Pizza Place  
1            Bikram Yoga       40.876844       -73.906204         Yoga Studio  
2          Tibbett Diner       40.880404       -73.908937               Diner  
3              Starbucks       40.877531       -73.905582         Coffee Shop  
4  Land & Sea Restaurant       40.877885       -73.905873  Seafood Restaurant

Let's check how many venues were returned for each neighborhood

In [50]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                      100                     100    100   
Carnegie Hill                          100                     100    100   
Central Harlem                          46                      46     46   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             43                      43     43   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                               100                     100    100   
Greenwich Village                      100                     100    100   
Hamilton Heights                        58                      58     58   
Hudson Yards                            59                      59     59   
Inwood                                  56                      56     56   
Lenox Hill                             100                     100    100   
Lincoln Square                         100                     100    100   
Little Italy                           100                     100    100   
Lower East Side                         62                      62     62   
Manhattan Valley                        64                      64     64   
Manhattanville                          38                      38     38   
Marble Hill                             24                      24     24   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     44                      44     44   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        25                      25     25   
Soho                                   100                     100    100   
Stuyvesant Town                         20                      20     20   
Sutton Place                           100                     100    100   
Tribeca                                100                     100    100   
Tudor City                              82                      82     82   
Turtle Bay                             100                     100    100   
Upper East Side                        100                     100    100   
Upper West Side                        100                     100    100   
Washington Heights                      87                      87     87   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City               100              100             100  
Carnegie Hill                   100              100             100  
Central Harlem                   46               46              46  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      43               43              43  
East Village                    100              100             100  
Finan

#### Let's find out how many unique categories can be curated from all the returned venues

In [51]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 330 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [52]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0                   0                    0               0             0   
1                   0                    0               0             0   
2                   0                    0               0             0   
3                   0                    0               0             0   
4                   0                    0               0             0   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0       0                 0                       0            0           0   
1       0                 0                       0            0           0   
2       0                 0                       0            0           0   
3       0                 0                       0            0           0   
4       0                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                      0                    0              0          0   
1                      0                    0              0          0   
2                      0                    0              0          0   
3                      0                    0              0          0   
4                      0                    0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       0     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Basketball Court  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0                 0                0         0            0           0   
1                 0                0         0            0           0   
2                 0                0         0            0           0   
3                 0                0         0            0           0   
4                 0                0         0            0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bike Shop  Bike Trail  \
0                   0                         0          0           0   
1                   0                         0          0           0   
2                   0                         0          0           0   
3                   0                         0          0           0   
4                   0                         0          0           0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  Boxing Gym  \
0       0           0              0          0         0           0   
1       0           0              0          0         0           0   
2       0           0              0          0         0           0   
3       0           0              0          0         0           0   
4       0           0              0          0         0           0   

   

And let's examine the new dataframe size.

In [53]:
manhattan_onehot.shape

(3308, 331)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [54]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.010000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.000000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.01               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.000000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.010000            0.00               0.00   
37   Washington Heights           0.011494            0.00               0.00   
38         West Village           0.010000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0             0.000000             0.010000            0.00          0.00   
1             0.000000             0.010000            0.00          0.00   
2             0.065217             0.043478            0.00          0.00   
3             0.000000             0.040000            0.00          0.00   
4             0.000000             0.040000            0.00          0.00   
5             0.000000             0.020000            0.00          0.01   
6             0.000000             0.040000            0.00          0.00   
7             0.000000             0.000000            0.00          0.00

#### Let's confirm the new size

In [55]:
manhattan_grouped.shape

(40, 331)

#### Let's print each neighborhood along with the top 5 most common venues

In [56]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
                venue  freq
0         Coffee Shop  0.08
1                Park  0.07
2               Hotel  0.05
3  Italian Restaurant  0.03
4           Wine Shop  0.03


----Carnegie Hill----
            venue  freq
0     Pizza Place  0.06
1     Coffee Shop  0.05
2  Cosmetics Shop  0.05
3            Café  0.04
4     Yoga Studio  0.03


----Central Harlem----
                  venue  freq
0    African Restaurant  0.07
1        Cosmetics Shop  0.04
2   American Restaurant  0.04
3  Gym / Fitness Center  0.04
4           Art Gallery  0.04


----Chelsea----
                 venue  freq
0          Coffee Shop  0.07
1   Italian Restaurant  0.06
2       Ice Cream Shop  0.05
3               Bakery  0.04
4  American Restaurant  0.04


----Chinatown----
                 venue  freq
0   Chinese Restaurant  0.09
1      Bubble Tea Shop  0.06
2  American Restaurant  0.04
3   Dim Sum Restaurant  0.04
4         Cocktail Bar  0.04


----Civic Center----
                  venue 

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [57]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [58]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City           Coffee Shop                  Park   
1      Carnegie Hill           Pizza Place           Coffee Shop   
2     Central Harlem    African Restaurant    Seafood Restaurant   
3            Chelsea           Coffee Shop    Italian Restaurant   
4          Chinatown    Chinese Restaurant       Bubble Tea Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Hotel             Wine Shop    Italian Restaurant   
1        Cosmetics Shop                  Café     French Restaurant   
2     French Restaurant        Cosmetics Shop           Art Gallery   
3        Ice Cream Shop                Bakery             Nightclub   
4          Cocktail Bar   American Restaurant    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Fountain         Memorial Site          Burger Joint   
1                    Bar             Bookstore                   Spa   
2     Chinese Restaurant  Gym / Fitness Center   American Restaurant   
3    American Restaurant               Theater                 Hotel   
4  Vietnamese Restaurant          Noodle House    Salon / Barbershop   

  9th Most Common Venue 10th Most Common Venue  
0            Food Truck             Food Court  
1           Yoga Studio    Japanese Restaurant  
2                   Spa   Ethiopian Restaurant  
3    Seafood Restaurant            Art Gallery  
4                Bakery      Hotpot Restaurant

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [59]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 4, 4, 2, 4, 2, 1, 4, 4, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [60]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               4   
2  Manhattan  Washington Heights  40.851903 -73.936900               4   
3  Manhattan              Inwood  40.867684 -73.921210               4   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               4   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Discount Store           Coffee Shop           Yoga Studio   
1    Chinese Restaurant       Bubble Tea Shop          Cocktail Bar   
2                  Café                Bakery     Mobile Phone Shop   
3                  Café    Mexican Restaurant           Pizza Place   
4    Mexican Restaurant         Deli / Bodega           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
0           Pizza Place       Supplement Shop             Steakhouse   
1   American Restaurant    Dim Sum Restaurant  Vietnamese Restaurant   
2        Sandwich Place           Supermarket                    Gym   
3                Lounge              Wine Bar             Restaurant   
4                  Café           Pizza Place           Liquor Store   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0         Shopping Mall            Shoe Store    Seafood Restaurant   
1          Noodle House    Salon / Barbershop                Bakery   
2    Chinese Restaurant            Shoe Store    Mexican Restaurant   
3                Bakery    Frozen Yogurt Shop         Deli / Bodega   
4          Cocktail Bar        Sandwich Place                School   

  10th Most Common Venue  
0         Sandwich Place  
1      Hotpot Restaurant  
2       Tapas Restaurant  
3     Chinese Restaurant  
4     Chinese Restaurant

Finally, let's visualize the resulting clusters

In [61]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [62]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
11     Roosevelt Island        Sandwich Place                  Park   
15              Midtown                 Hotel            Food Truck   
26  Morningside Heights           Coffee Shop            Food Truck   
28    Battery Park City           Coffee Shop                  Park   
29   Financial District           Coffee Shop                 Hotel   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11                   Gym        Farmers Market           Dry Cleaner   
15            Steakhouse               Theater           Coffee Shop   
26                  Park   American Restaurant             Bookstore   
28                 Hotel             Wine Shop    Italian Restaurant   
29            Steakhouse             Wine Shop                   Gym   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11                School        Scenic Lookout      Greek Restaurant   
15                   Spa          Cocktail Bar   American Restaurant   
26          Burger Joint         Deli / Bodega          Tennis Court   
28              Fountain         Memorial Site          Burger Joint   
29                   Bar            Food Truck           Pizza Place   

   9th Most Common Venue   10th Most Common Venue  
11          Liquor Store                    Train  
15        Clothing Store                Bookstore  
26        Sandwich Place  New American Restaurant  
28            Food Truck               Food Court  
29    Italian Restaurant                Juice Bar

#### Cluster 2

In [63]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
13  Lincoln Square               Theater  Gym / Fitness Center   
14         Clinton               Theater           Coffee Shop   
39    Hudson Yards           Coffee Shop    Italian Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
13                 Plaza    Italian Restaurant          Concert Hall   
14    Italian Restaurant   American Restaurant  Gym / Fitness Center   
39                  Café  Gym / Fitness Center            Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13     French Restaurant                  Café           Opera House   
14                   Spa                 Hotel             Wine Shop   
39               Theater                 Hotel   American Restaurant   

    9th Most Common Venue 10th Most Common Venue  
13  Performing Arts Venue                   Park  
14                    Gym                Dog Run  
39                    Gym            Art Gallery

#### Cluster 3

In [64]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
5      Manhattanville         Deli / Bodega    Seafood Restaurant   
8     Upper East Side    Italian Restaurant               Exhibit   
9           Yorkville           Coffee Shop                   Bar   
10         Lenox Hill    Italian Restaurant           Coffee Shop   
12    Upper West Side    Italian Restaurant                   Bar   
16        Murray Hill           Coffee Shop                 Hotel   
17            Chelsea           Coffee Shop    Italian Restaurant   
18  Greenwich Village    Italian Restaurant      Sushi Restaurant   
21            Tribeca    Italian Restaurant                  Park   
23               Soho        Clothing Store              Boutique   
24       West Village    Italian Restaurant        Cosmetics Shop   
27           Gramercy    Italian Restaurant           Coffee Shop   
31               Noho    Italian Restaurant     French Restaurant   
32       Civic Center  Gym / Fitness Center    Italian Restaurant   
33      Midtown South     Korean Restaurant                 Hotel   
34       Sutton Place  Gym / Fitness Center    Italian Restaurant   
35         Turtle Bay    Italian Restaurant      Sushi Restaurant   
38           Flatiron    Italian Restaurant           Yoga Studio   

      3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
5        Italian Restaurant      Mexican Restaurant             BBQ Joint   
8               Art Gallery             Coffee Shop             Juice Bar   
9                       Gym      Italian Restaurant           Pizza Place   
10         Sushi Restaurant             Pizza Place  Gym / Fitness Center   
12              Coffee Shop                Wine Bar                Bakery   
16       Italian Restaurant     Japanese Restaurant        Sandwich Place   
17           Ice Cream Shop                  Bakery             Nightclub   
18        French Restaurant          Clothing Store     Indian Restaurant   
21      American Restaurant                     Spa                  Café   
23            Women's Store             Men's Store            Shoe Store   
24  New American Restaurant               Jazz Club             Gastropub   
27             Cocktail Bar               Wine Shop            Restaurant   
31              Coffee Shop            Cocktail Bar                 Hotel   
32                   Bakery       French Restaurant           Yoga Studio   
33                Hotel Bar          Cosmetics Shop           Coffee Shop   
34        Indian Restaurant  Furniture / Home Store             Juice Bar   
35                    Hotel             Coffee Shop              Wine Bar   
38     Gym / Fitness Center     American Restaurant   Sporting Goods Shop   

   6th Most Common Venue     7th Most Common Venue  \
5            Beer Garden         Food & Drink Shop   
8                 Bakery      Gym / Fitness Center   
9     Mexican Restaurant             Deli / Bodega   
10   Sporting Goods Shop              Burger Joint   
12          Burger Joint         Indian Restaurant   
16                   Spa                       Gym   
17   American Restaurant                   Theater   
18                  Café        Seafood Restaurant   
21              Boutique                  Wine Bar   
23           Art Gallery  Mediterranean Restaurant   
24              Wine Bar                    Bakery   
27            Bagel Shop        Mexican Restaurant   
31             Bookstore        Mexican Restaurant   
32           Coffee Shop              Cocktail Bar   
33   Japanese Restaurant        Italian Restaurant   
34          Dessert Shop       American Restaurant   
35                  Park                Steakhouse   
38    Salon / Barbershop   New American Restaurant   

            8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                            Park          Burger Joint                    Bar  
8               French Restaurant           

#### Cluster 4

In [65]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                   Bar                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37            Playground         Boat or Ferry      Basketball Court   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37              Heliport        Baseball Field           Gas Station   

   9th Most Common Venue 10th Most Common Venue  
37           Beer Garden        Harbor / Marina

#### Cluster 5

In [66]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0          Marble Hill        Discount Store           Coffee Shop   
1            Chinatown    Chinese Restaurant       Bubble Tea Shop   
2   Washington Heights                  Café                Bakery   
3               Inwood                  Café    Mexican Restaurant   
4     Hamilton Heights    Mexican Restaurant         Deli / Bodega   
6       Central Harlem    African Restaurant    Seafood Restaurant   
7          East Harlem    Mexican Restaurant                Bakery   
19        East Village                   Bar        Ice Cream Shop   
20     Lower East Side                  Café           Coffee Shop   
22        Little Italy                Bakery                  Café   
25    Manhattan Valley           Coffee Shop           Pizza Place   
30       Carnegie Hill           Pizza Place           Coffee Shop   
36          Tudor City    Mexican Restaurant                  Park   

   3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0            Yoga Studio                Pizza Place       Supplement Shop   
1           Cocktail Bar        American Restaurant    Dim Sum Restaurant   
2      Mobile Phone Shop             Sandwich Place           Supermarket   
3            Pizza Place                     Lounge              Wine Bar   
4            Coffee Shop                       Café           Pizza Place   
6      French Restaurant             Cosmetics Shop           Art Gallery   
7          Deli / Bodega  Latin American Restaurant       Thai Restaurant   
19              Wine Bar               Cocktail Bar    Mexican Restaurant   
20           Art Gallery           Ramen Restaurant        Sandwich Place   
22           Yoga Studio               Cocktail Bar       Bubble Tea Shop   
25           Yoga Studio         Mexican Restaurant                   Spa   
30        Cosmetics Shop                       Café     French Restaurant   
36      Greek Restaurant                Pizza Place                  Café   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Steakhouse         Shopping Mall            Shoe Store   
1   Vietnamese Restaurant          Noodle House    Salon / Barbershop   
2                     Gym    Chinese Restaurant            Shoe Store   
3              Restaurant                Bakery    Frozen Yogurt Shop   
4            Liquor Store          Cocktail Bar        Sandwich Place   
6      Chinese Restaurant  Gym / Fitness Center   American Restaurant   
7                    Café              Beer Bar            Taco Place   
19     Chinese Restaurant           Pizza Place      Ramen Restaurant   
20                   Park                Bakery          Cocktail Bar   
22     Seafood Restaurant        Sandwich Place    Salon / Barbershop   
25                   Café       Thai Restaurant                   Bar   
30                    Bar             Bookstore                   Spa   
36                Dog Run                 Diner                 Hotel   

   9th Most Common Venue     10th Most Common Venue  
0     Seafood Restaurant             Sandwich Place  
1                 Bakery          Hotpot Restaurant  
2     Mexican Restaurant           Tapas Restaurant  
3          Deli / Bodega         Chinese Restaurant  
4                 School         Chinese Restaurant  
6                    Spa       Ethiopian Restaurant  
7      French Restaurant                 Street Art  
19             Speakeasy                Record Shop  
20           Pizza Place  Latin American Restaurant  
22    Chinese Restaurant             Ice Cream Shop  
25            Playground              Deli / Bodega  
30           Yoga Studio        Japanese Restaurant  
36         Deli / Bodega                        Spa

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/) and [Polong Lin](https://www.linkedin.com/in/polonglin/). I hope you found this lab interesting and educational. Feel free to contact us if you have any questions!

This notebook is part of a course on **Coursera** called *Applied Data Science Capstone*. If you accessed this notebook outside the course, you can take this course online by clicking [here](http://cocl.us/DP0701EN_Coursera_Week3_LAB2).

<hr>

Copyright &copy; 2018 [Cognitive Class](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).